src/java/org/apache/cassandra/cql3/restrictions/StatementRestrictions.java

```diff
@@ -302,11 +305,24 @@ public StatementRestrictions build()
             if (allowUseOfSecondaryIndices && type.allowUseOfSecondaryIndices())
                 indexRegistry = IndexRegistry.obtain(table);
 
-            return doBuild(whereClause.root(), indexRegistry);
+
+            WhereClause.AndElement root = whereClause.root().conjunctiveForm();
+            return doBuild(root, indexRegistry, 0);
         }
 
-        StatementRestrictions doBuild(WhereClause.ExpressionElement element, IndexRegistry indexRegistry)
+        /**
+         * Processes the WHERE clause expression tree recursively and assigns the restrictions to different sets
+         * based on the columns they are applied to.
+         *
+         * @param element root of the tree
+         * @param nestingLevel recursion depth needed to reject the restrictions that
+         *                     are not allowed to be nested (e.g. partition key restrictions)
+         */
+        StatementRestrictions doBuild(WhereClause.ExpressionElement element, IndexRegistry indexRegistry, int nestingLevel)
         {
+            assert element instanceof WhereClause.AndElement || nestingLevel > 0:
```

Comment:
I feel it would be better if we added an `isConjunctive` method to the `ExpressionElement`. We can then allow the `EmptyElement` to return this as `true` and also result itself from `conjunctiveForm`. At present we are creating a new `AndElement` for every query with an empty where clause.

We can then use that method here rather than an `instanceof`.

Reply:
I'm generally not a big fan of replacing an `instanceof` by polymorphism. I find polymorphism has actually bigger cognitive load here - one must inspect all the classes in the hierarchy to fully understand what's going on, and what `isConjunctive` really means (check which ones return true and which ones return false), to come to the conclusion that "oh yeah, so you're just checking if it is an And node...". 

What I like in the current `conjunctiveForm` is the fact it already returns an `AndElement` and you would not need any further instanceof or downcasting if only there was no recursion in the `doBuild` method. 

However, your point about wrapping an `EmptyElement` into `AndElement` stands. That's not elegant now, because `conjunctiveForm` does not preserve "emptiness". So, what about just dropping the `EmptyElement` class and replacing the empty value by an empty `AndElement`?

Reply:
On more thing: What would the empty nodes evaluate to? True or false? How would an expressions like this print (Empty AND Empty)? Looks like a very special case we have only to allow no WHERE clause. But semantically a lack of the where clause is equivalent to `WHERE TRUE`. 

If we insist on the expression elements to allow "empty" values, I think that printing empty `AndElement` as `"TRUE"` and and empty `OrElement` as `"FALSE"` would be a lot more elegant, although they would not parse back (but `()`, or `""` which were used before don't parse either).

Reply:
> However, your point about wrapping an EmptyElement into AndElement stands. That's not elegant now, because conjunctiveForm does not preserve "emptiness". So, what about just dropping the EmptyElement class and replacing the empty value by an empty AndElement?

That works for me. I was more concerned about having unnecessary object creation.

>On more thing: What would the empty nodes evaluate to? True or false? How would an expressions like this print (Empty AND Empty)? Looks like a very special case we have only to allow no WHERE clause. But semantically a lack of the where clause is equivalent to WHERE TRUE.

This does seem to be another CQL specific thing. I currently doesn't allow `WHERE` with nothing following it. So we really ought to not allow `WHERE ()`. I just feel that this is a change that we don't need. 

Also, I can understand that `WHERE TRUE` would just be the equivalent of having no where clause but what would expect `WHERE FALSE` to do? (just for interest)